src/emulator/taskQueue.ts

```diff
@@ -0,0 +1,380 @@
+      const currentFirst = this.first;
+      this.first = this.first.next;
+      if (this.last === currentFirst) {
+        this.last = null;
+      }
+      this.count--;
+      return currentFirst.data;
+    } else {
+      throw new Error("Trying to dequeue from an empty queue");
+    }
+  }
+
+  remove(id: string): void {
+    if (this.nodeMap[id] === undefined) {
+      throw new Error("Trying to remove a task that doesn't exist");
+    }
+    const toRemove = this.nodeMap[id];
+
+    if (toRemove.next === null && toRemove.prev === null) {
+      this.first = null;
+      this.last = null;
+    } else if (toRemove.next === null) {
+      this.last = toRemove.prev;
+      toRemove.prev!.next = null;
+    } else if (toRemove.prev === null) {
+      this.first = toRemove.next;
+      toRemove.next.prev = null;
+    } else {
+      const prev = toRemove.prev;
+      const next = toRemove.next;
+      prev.next = next;
+      next.prev = prev;
+    }
+    delete this.nodeMap[id];
+    this.count--;
+  }
+
+  getAll(): T[] {
+    const all = [];
+    let curr = this.first;
+    while (curr) {
+      all.push(curr.data);
+      curr = curr.next;
+    }
+    return all;
+  }
+
+  isEmpty(): boolean {
+    return this.first === null;
+  }
+
+  size(): number {
+    return this.count;
+  }
+}
+
+export enum TaskStatus {
+  // When a task has been created/retried and should check if it should run
+  NOT_STARTED,
+  // When a task has been dispatched and is currently running
+  RUNNING,
+  // When a task has been dispatched and failed, but will be tried again
+  RETRY,
+  // When a task has failed and exhausted it's retry parameters and will not be tried again.
+  FAILED,
+  // When a task has been completed successfully
+  FINISHED,
+}
+
+export interface EmulatedTaskMetadata {
+  currentAttempt: number;
+  currentBackoff: number;
+  startTime: number;
+  status: TaskStatus;
+  lastRunTime: number | null;
+}
+
+export interface EmulatedTask {
+  task: Task;
+  metadata: EmulatedTaskMetadata;
+}
+
+export class TaskQueue {
+  queue: Queue<EmulatedTask> = new Queue<EmulatedTask>();
+  logger = EmulatorLogger.forEmulator(Emulators.TASKS);
+  static TASK_QUEUE_INTERVAL = 1000;
+
+  // Current number of tokens the queue has
+  private tokens = 0;
+  // The maximum number of tokens that can fit in the "bucket"
+  private maxTokens;
+  // The last time the token bucket was updated, used in calculations of how many tokens to add.
+  private lastTokenUpdate;
+  // The IDs of all the tasks ever queued in this session to allow for deduplication
+  private queuedIds: Set<string>;
+  // The tasks that have been dispatched that the queue is waiting on
+  private dispatches: (EmulatedTask | null)[];
+  // The indexes of the open slots in the dispatch array
+  private openDispatches: number[];
+
+  constructor(
+    private key: string,
+    private config: TaskQueueConfig,
+  ) {
+    this.maxTokens = Math.max(this.config.rateLimits.maxDispatchesPerSecond, 1.1);
```

Comment:
So I initially set this to a min  of 1 (if the  max is less than 1 then no tasks will ever be dispatched). 

The issue with this was that I was testing out a rate of 0.2 per second (1 task every 5 seconds), tokens would be added ~ every 0.2 seconds, however due to imprecise timing, it would often add 0.19...

So after 5 seconds, the token bucket would have  have ~0.99... tokens in it, and would not dispatch a task. It would take an additional update cycle (1 second) to update the tokens and then dispatch a task, and when the task got dispatched it would reset to 0, causing the rate to be closer to 1 task every 6 seconds instead of every 5.

This way, the first task might still take ~6 seconds to run, but subsequent tasks should run closer to every 5 seconds. 

There are  probably other ways to resolve this including: 
- Keep at 1 but increase resolution for refilling the bucket
- Add tokens in discrete amounts (would still be difficult to handle rates less than 1)